SerialPrograms/Source/PokemonSV/Programs/AutoStory/PokemonSV_MenuOption.cpp

```diff
@@ -0,0 +1,214 @@
+/*  Menu Option Session
+ *
+ *  From: https://github.com/PokemonAutomation/Arduino-Source
+ *
+ */
+
+#include <algorithm>
+#include "Common/Cpp/Containers/FixedLimitVector.tpp"
+#include "Common/Cpp/Concurrency/AsyncDispatcher.h"
+#include "NintendoSwitch/Commands/NintendoSwitch_Commands_PushButtons.h"
+#include "CommonFramework/Exceptions/OperationFailedException.h"
+#include "CommonFramework/VideoPipeline/VideoFeed.h"
+#include "CommonFramework/Tools/ConsoleHandle.h"
+#include "CommonFramework/OCR/OCR_NumberReader.h"
+#include "PokemonSV_MenuOption.h"
+#include "PokemonSV/Inference/PokemonSV_MenuOptionReader.h"
+
+#include "Common/Cpp/PrettyPrint.h"
+#include <iostream>
+// using std::cout;
+// using std::endl;
+
+namespace PokemonAutomation{
+namespace NintendoSwitch{
+namespace PokemonSV{
+
+MenuOption::~MenuOption() = default;
+
+MenuOption::MenuOption(
+    ConsoleHandle& console, 
+    BotBaseContext& context,
+    Language language
+)
+    : m_console(console)
+    , m_context(context)
+    , m_language(language)
+    , m_overlays(console.overlay())
+    , m_arrow(COLOR_CYAN, GradientArrowType::RIGHT, {0.02, 0.10, 0.05, 0.80})
+{
+    // reader.make_overlays(m_overlays);
+    for (size_t c = 0; c < 10; c++){
+        m_boxes_item[c] = ImageFloatBox(0.055, 0.135 + c * 0.0739, 0.260, 0.060);
+        m_boxes_toggle[c] = ImageFloatBox(0.345, 0.135 + c * 0.0739, 0.175, 0.060);
+        m_overlays.add(COLOR_BLUE, m_boxes_item[c]);
+        m_overlays.add(COLOR_BLUE, m_boxes_toggle[c]);
+    }
+}
+
+void MenuOption::set_options(
+    std::vector<std::pair<MenuOptionItemEnum, std::vector<MenuOptionToggleEnum>>> options
```

Comment:
I think there's several more places like this where you're copying a non-trivial parameter instead of passing by const-ref.

Reply:
Basic programming question: For the vector of MenuOptionToggleEnum, within the pair, do I need to also make this a reference as well? Or is this redundant since I'm already making `options` pass by reference?

i.e.
`const std::vector<std::pair<MenuOptionItemEnum, std::vector<MenuOptionToggleEnum>&>>& options`
vs
`const std::vector<std::pair<MenuOptionItemEnum, std::vector<MenuOptionToggleEnum>>>& options`

Reply:
The general rule of thumb for passing in read-only parameters is:

- If it's a primitive type (integer, float/double, pointer, enum): Pass by value. You don't need to mark it const because it's being copied and thus the function cannot affect the caller.
- If it's anything else, (struct, class, array, etc...): Pass by const reference. This avoids the copying the entire object since passing a reference is (internally), the same a passing a single pointer.

Of course exceptions will apply. Especially with async stuff where you have to watch out for the lifetimes of the objects.